libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Generators/ActivityChecker.cs

```diff
@@ -0,0 +1,372 @@
+                if (action is JValue jValue && jValue.Type == JTokenType.String)
+                {
+                    return result;
+                }
+                else if (action is JObject actionJObj)
+                {
+                    result.AddRange(CheckCardAction(actionJObj));
+                }
+            }
+
+            return result;
+        }
+
+        private static List<Diagnostic> CheckCardAction(JObject cardActionJObj)
+        {
+            var result = new List<Diagnostic>();
+            var type = GetStructureType(cardActionJObj);
+            if (type != nameof(CardAction).ToLower())
+            {
+                result.Add(BuildDiagnostic($"'{type}' is not card action type.", false));
+            }
+            else
+            {
+                foreach (var item in cardActionJObj)
+                {
+                    var property = item.Key.Trim();
+                    var value = item.Value.ToString().ToLower();
+
+                    switch (property.ToLower())
+                    {
+                        case "$type":
+                        case "title":
+                        case "value":
+                        case "displaytext":
+                        case "text":
+                        case "image":
+                            break;
+
+                        case "type":
+                            if (value != ActionTypes.ImBack.ToLower()
+                                && value != ActionTypes.Call.ToLower()
+                                && value != ActionTypes.DownloadFile.ToLower()
+                                && value != ActionTypes.MessageBack.ToLower()
+                                && value != ActionTypes.openApp.ToLower()
+                                && value != ActionTypes.OpenUrl.ToLower()
+                                && value != ActionTypes.Payment.ToLower()
+                                && value != ActionTypes.PlayAudio.ToLower()
+                                && value != ActionTypes.PlayVideo.ToLower()
+                                && value != ActionTypes.PostBack.ToLower()
+                                && value != ActionTypes.ShowImage.ToLower()
+                                && value != ActionTypes.Signin.ToLower())
+                            {
+                                result.Add(BuildDiagnostic($"'{value}' is not a valid action type"));
+                            }
+
+                            break;
+                        default:
+                            result.Add(BuildDiagnostic($"'{property}' is not support for card action.", false));
+                            break;
+                    }
+                }
+            }
+
+            return result;
+        }
+
+        private static string GetStructureType(JObject jObj)
+        {
+            if (jObj == null)
+            {
+                return string.Empty;
+            }
+
+            var type = jObj["$type"]?.ToString()?.Trim();
```

Comment:
Could this code be cleaned up so the null programming can be removed?

Reply:
what is the right way to handler the null programming?